### DISCUSSION TITLE
How can I deliver user data to the implemented function?

### DISCUSSION THREAD

**Discussion by ytrewq271828** [ANLAB-KAIST/KENSv3#156] [Project Q&A 2. Transport Layer (TCP)] [answered]
Hello. I wonder how can I get the application-layer data in KENS.
I was working on implementing the state transition diagram.
But I was not able to find any documentation regarding rdt_send(data) function.
I think somehow the application layer data should be delivered to transport layer.
Are there any specific parameter or predefined methods about the application data?

Thanks in advance.

**Comment by Pusnow** [comment, marked as answer]
In the `systemCallBack()` method, you can retrieve application data by reading `SystemCallParameter `. For example, for `rdt_send`, you can use `write()` system call's `params[1]` for pointer to application data and `params[2]` for the data's length.

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/app/kens/TCPAssignment.cpp#L51-L52


BTW, you don't need to implement data transfer in PA2. It is in PA3.
